torch/lowrank.py

```diff
@@ -0,0 +1,325 @@
+    a centered matrix :attr:`A` such that :math:`A = U diag(S) V^T`.
+
+    .. note:: Different from the standard SVD, the size of returned
+              matrices depend on the specified rank and q
+              values as follows:
+                - U is m x q matrix
+                - S is q-vector
+                - V is n x q matrix
+
+    .. note:: To obtain repeatable results, reset the seed for the
+              pseudorandom number generator
+
+    Arguments:
+
+        A (Tensor): the input tensor of size :math:`(*, m, n)`
+
+        q (int, optional): a slightly overestimated rank of A. By
+                           default, q = min(6, m, n).
+
+        center (bool, optional): if True, center the input tensor,
+                                 otherwise, assume that the input is
+                                 centered.
+
+        niter (int, optional): the number of subspace iterations to
+                               conduct; niter must be a nonnegative
+                               integer, and defaults to 2
+
+    References::
+
+        - Nathan Halko, Per-Gunnar Martinsson, and Joel Tropp, Finding
+          structure with randomness: probabilistic algorithms for
+          constructing approximate matrix decompositions,
+          arXiv:0909.4061 [math.NA; math.PR], 2009 (available at
+          `arXiv <http://arxiv.org/abs/0909.4061>`_).
+
+    """
+    (m, n) = A.shape[-2:]
+
+    if q is None:
+        q = min(6, m, n)
+    elif not (q >= 0 and q <= min(m, n)):
+        raise ValueError('q(={}) must be non-negative integer'
+                         ' and not greater than min(m, n)={}'
+                         .format(q, min(m, n)))
+    if not (niter >= 0):
+        raise ValueError('niter(={}) must be non-negative integer'
+                         .format(niter))
+
+    dtype = get_floating_dtype(A)
+
+    if not center:
+        return svd(A, q, niter=niter)
+
+    if is_sparse(A):
+        if len(A.shape) != 2:
+            raise ValueError('pca input is expected to be 2-dimensional tensor')
+        c = torch.sparse.sum(A, dim=-2) / m
+        # reshape c
+        column_indices = c.indices()[0]
+        indices = torch.zeros(2, len(column_indices),
+                              dtype=column_indices.dtype,
+                              device=column_indices.device)
+        indices[0] = column_indices
+        C_t = torch.sparse_coo_tensor(
+            indices, c.values(), (n, 1), dtype=dtype, device=A.device)
+
+        ones_m1_t = torch.ones(A.shape[:-2] + (1, m), dtype=dtype, device=A.device)
+        M = batch_transpose(torch.sparse.mm(C_t, ones_m1_t))
+        return svd(A, q, niter=niter, M=M)
+    else:
+        c = A.sum(axis=-2) / m
+        C = c.reshape(A.shape[:-2] + (1, n))
+        ones_m1 = torch.ones(A.shape[:-1] + (1, ), dtype=dtype, device=A.device)
+        M = ones_m1.matmul(C)
+        return svd(A - M, q, niter=niter)
```

Comment:
What's the difference between `A-M` here and line 319: `svd(A, M)`?

Reply:
When `A` and `M` are dense tensors, then one can use `A-M` as it will save computations.

When either `A` or `M` is a sparse tensor (presumably large), using `A-M` may fail due to limited resources.

Reply:
If A and M are dense matrices, caling `A-M` still form a third new dense matrix, and so there would still be advantages in not forming the difference matrix, no?

Reply:
This is true only if the system is short of memory when holding dense `A` and `M`.
When the system can hold `A - M`, then CPU time will be saved more than 2x (e.g. consider computing `AR - MR` vs `(AM)*R` where the former expression contains two matrix products and elementwise addition wheil the later contains only one matix product).

Reply:
That also holds for sparse matrices, right?

Reply:
yes, so there are two algorithms for low-rank PCA:
1. one that is faster but consumes more memory: `svd(A - M)`
2. another that is memory efficient but does more computations: `svd(A, M)`.
For cases when there are plenty of memory resources available, the first algorithm is preferred for both sparse and dense inputs. Otherwise, the second.

That said, note that using the low-rank matrix algorithms makes sense only for really large matrices when low-rank can be presumed. This is usually the case in PCA related applications.
In all other cases, that is when the input matrices are full-rank or small, one should not use a low-rank SVD algorithm because the `torch.svd` will be much more efficient or the low-rank assumption is violated.

Hence, in the context of this PR that implements the low-rank matrix algorithms, the inputs are presumed to be large sparse matrices that `torch.svd` cannot handle. These algorithms can be used for smaller dense matrices but it would be a waste of computational resources and not recommended.

Reply:
> yes, so there are two algorithms for low-rank PCA:
> 
> 1. one that is faster but consumes more memory: `svd(A - M)`
> 2. another that is memory efficient but does more computations: `svd(A, M)`.
> 
> For cases when there are plenty of memory resources available, the first algorithm is preferred for both sparse and dense inputs. Otherwise, the second.

Thanks for the details! So the assumption in the code is that someone uses sparse matrices if and only if memory is an issue. This is good enough for the scope of this PR. We could keep that in mind as an option to expose to the user later.

Reply:
I have forgotten to mention another important point: the shape of `M` is `(1, n)` because `M` contains the mean of `A` columns. So, using the `svd(A-M)` variant means that `M` must be expanded to the shape of `A` that would lead to a considerable increase of memory as well as processor usage. This fact actually makes the usage of `svd(A, M)` variant even more advantageous.